web/packages/teleport/src/Navigation/Navigation.tsx

```diff
@@ -105,32 +105,55 @@ export function Navigation() {
       NavigationCategory.Resources
   );
 
-  const [previousRoute, setPreviousRoute] = useState('');
+  const [previousRoute, setPreviousRoute] = useState<{ [key: string]: string }>(
+    {}
+  );
 
-  useEffect(() => {
-    return history.listen(next => {
+  const handleLocationChange = useCallback(
+    (next: Location<unknown>) => {
       const previousPathName = location.pathname;
 
       const category = getCategoryForRoute(features, next);
 
       if (category && category !== view) {
-        setPreviousRoute(previousPathName);
+        setPreviousRoute(previous => ({
+          ...previous,
+          [view]: previousPathName,
+        }));
         setView(category);
       }
-    });
+    },
+    [location, view]
+  );
+
+  useEffect(() => {
+    return history.listen(handleLocationChange);
   }, [history, location.pathname, features, view]);
 
+  const handlePopState = useCallback(
+    (event: PopStateEvent) => {
+      handleLocationChange((event.currentTarget as any).location);
```

Comment:
What would you say about this:

<details>
<summary>Proposed changes</summary>

```diff
diff --git a/web/packages/teleport/src/Navigation/Navigation.tsx b/web/packages/teleport/src/Navigation/Navigation.tsx
index 960eb3f44b..b1ad0d5f7a 100644
--- a/web/packages/teleport/src/Navigation/Navigation.tsx
+++ b/web/packages/teleport/src/Navigation/Navigation.tsx
@@ -33,7 +33,7 @@ import { NavigationCategoryContainer } from 'teleport/Navigation/NavigationCateg
 
 import logo from './logo.png';
 
-import type { Location } from 'history';
+import type * as history from 'history';
 
 import type { TeleportFeature } from 'teleport/types';
 
@@ -77,7 +77,7 @@ export function getFirstRouteForCategory(
 
 function getCategoryForRoute(
   features: TeleportFeature[],
-  route: Location<unknown>
+  route: history.Location<unknown> | Location
 ) {
   const feature = features
     .filter(feature => Boolean(feature.route))
@@ -110,7 +110,7 @@ export function Navigation() {
   );
 
   const handleLocationChange = useCallback(
-    (next: Location<unknown>) => {
+    (next: history.Location<unknown> | Location) => {
       const previousPathName = location.pathname;
 
       const category = getCategoryForRoute(features, next);
@@ -132,7 +132,7 @@ export function Navigation() {
 
   const handlePopState = useCallback(
     (event: PopStateEvent) => {
-      handleLocationChange((event.currentTarget as any).location);
+      handleLocationChange((event.currentTarget as Window).location);
     },
     [view]
   );
```
</details>